### PR TITLE
Row by row diffs

### LINKED ISSUE #76: Better display for matrices
``` r
x <- y <- matrix(1:4, nrow = 2)
y[2, 2] <- 5L
waldo::compare(x, y)
#> `old`: 1 2 3 4
#> `new`: 1 2 3 5
```

<sup>Created on 2021-04-19 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

### LINKED ISSUE #78: Better display for data.frames
Related to #76, it would be nice to have a more natural view of where things differ in data.frames and tibbles.

```
> iris_mod <- iris
> iris_mod[3,2] <- 17
> compare(iris, iris_mod)
`old$Sepal.Width[1:6]`: 4 3  3 3 4 4
`new$Sepal.Width[1:6]`: 4 3 17 3 4 4
```

### PR DESCRIPTION
**Pull request by hadley**
Uses printed representation of object, so must always be accompanied by fallback

Fixes #76. Fixes #78.

### CODE REVIEW

**Comment by hadley** [comment]
@jennybc no need to look at the code because it's probably hard to follow, but I'd appreciate an extra pair of eyes on the [snasphot results](https://github.com/r-lib/waldo/pull/90/files#diff-8b436012a9fdcd55e60186939aa89c03d58b688376cf37442ca23fd2cba646fc), particularly with an eye to thinking about important cases that I should cover. This is just a first stab at this problem so it doesn't need to be super comprehensive, but I do want to make it more helpful than the existing output.

File: tests/testthat/_snaps/compare-data-frame.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+      `attr(old, 'row.names')`: 1 2 3 4 5
+      `attr(new, 'row.names')`: 1 2 3    
+      
+      old vs new
+              x y
+        [1, ] 1 5
+        [2, ] 2 4
+        [3, ] 3 3
+      - [4, ] 4 2
+      - [5, ] 5 1
```
Comment (jennybc) [inline tests/testthat/_snaps/compare-data-frame.md:16]: This feels a bit weird. In this case, the `old$x` vs. `new$x` comparisons definitely feel easier to interpret. From the above, I feel unsure who is missing these rows (I mean, I know, but it's not super obvious from what I'm looking at).

File: tests/testthat/_snaps/compare-data-frame.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      
+      `old$x`: 1 2 3 4 5
+      `new$x`: 1 2 3    
+      
+      `old$y`: 5 4 3 2 1
+      `new$y`: 5 4 3
```
Comment (jennybc) [inline tests/testthat/_snaps/compare-data-frame.md:22]: Off topic: feels like there should be an empty line here.

File: tests/testthat/_snaps/compare-data-frame.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    Code
+      compare(df, unrowname(df[c(1, 5, 2, 3, 4, 5), ]))
+    Output
+      `attr(old, 'row.names')[3:5]`: 3 4 5  
+      `attr(new, 'row.names')[3:6]`: 3 4 5 6
+      
+      old vs new
+              x y
+        [1, ] 1 5
+      + [2, ] 5 1
```
Comment (jennybc) [inline tests/testthat/_snaps/compare-data-frame.md:32]: Similar to the previous example, it's not super obvious that `new` has an insertion.

File: tests/testthat/_snaps/compare-data-frame.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    Code
+      df1 <- data.frame(x = 1:3, y = 1, z = c("a", "b", "c"))
+      df2 <- data.frame(x = c(1, 100, 3), y = 1, z = c("a", "B", "c"))
+      compare(df1, df2)
+    Output
+      old vs new
+                x z
+        [1, ]   1 a
+      - [2, ]   2 b
+      + [2, ] 100 B
```
Comment (jennybc) [inline tests/testthat/_snaps/compare-data-frame.md:54]: I continue to not love this, but nor do I have a concrete counterproposal.

File: tests/testthat/_snaps/compare-data-frame.md (hunk 1 above)
Comment (hadley) [reply tests/testthat/_snaps/compare-data-frame.md:16]: How about this?

```
old vs new
           x y
  old[1, ] 1 5
  old[2, ] 2 4
  old[3, ] 3 3
- old[4, ] 4 2
- old[5, ] 5 1
```

File: tests/testthat/_snaps/compare-data-frame.md (hunk 3 above)
Comment (hadley) [reply tests/testthat/_snaps/compare-data-frame.md:32]: And this:

```
             x z
  old[1, ]   1 a
- old[2, ]   2 b
+ new[2, ] 100 B
  old[3, ]   3 c
```

File: tests/testthat/_snaps/compare-data-frame.md (hunk 1 above)
Comment (jennybc) [reply tests/testthat/_snaps/compare-data-frame.md:16]: That feels worse 😂 because now I'm pretty sure that `old` is missing those rows (except the truth is that they're missing from `new`).

File: tests/testthat/_snaps/compare-data-frame.md (hunk 3 above)
Comment (jennybc) [reply tests/testthat/_snaps/compare-data-frame.md:32]: This helps (the explicit use of `old` and `new`.

File: tests/testthat/_snaps/compare-data-frame.md (hunk 1 above)
Comment (hadley) [reply tests/testthat/_snaps/compare-data-frame.md:16]: It's telling you the changes from `old` you need to make to get `new`; that at least is consistent across all waldo output.
